include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h

```diff
@@ -0,0 +1,383 @@
+                                                  __iters_per_work_item>{__reduce_op, _TransformOp{__transform_op}};
+        auto __reduce_pattern = unseq_backend::reduce_over_group<_ExecutionPolicy, _ReduceOp, _Tp>{__reduce_op};
+
+        const _Size __n_items = __ceiling_div(__n, __iters_per_work_item); // number of work items
+
+        sycl::buffer<_Tp> __res(sycl::range<1>(1));
+
+        sycl::event __reduce_event = __exec.queue().submit([&, __n, __n_items](sycl::handler& __cgh) {
+            oneapi::dpl::__ranges::__require_access(__cgh, __rngs...); // get an access to data under SYCL buffer
+            auto __res_acc = __res.template get_access<access_mode::write>(__cgh);
+            __dpl_sycl::__local_accessor<_Tp> __temp_local(sycl::range<1>(__work_group_size), __cgh);
+            __cgh.parallel_for<_ReduceKernel>(
+                sycl::nd_range<1>(sycl::range<1>(__work_group_size), sycl::range<1>(__work_group_size)),
+                [=](sycl::nd_item<1> __item_id) {
+                    ::std::size_t __global_idx = __item_id.get_global_id(0);
+                    ::std::uint16_t __local_idx = __item_id.get_local_id(0);
+                    // 1. Initialization (transform part). Fill local memory
+                    __transform_pattern.parallel(__local_idx, __n, __iters_per_work_item, __global_idx,
+                                                 /*global_offset*/ 0, __temp_local, __rngs...);
+                    __dpl_sycl::__group_barrier(__item_id);
+                    // 2. Reduce within work group using local memory
+                    _Tp __result = __reduce_pattern(__item_id, __global_idx, __n_items, __temp_local);
+                    if (__local_idx == 0)
+                    {
+                        __reduce_pattern.apply_init(__init, __result);
+                        __res_acc[0] = __result;
+                    }
+                });
+        });
+
+        return __future(__reduce_event, __res);
+    }
+}; // struct __parallel_transform_reduce_small_submitter
+
+// General parallel_transform_reduce - uses a tree-based reduction
+template <typename _Tp, typename _IsGPU>
+struct __parallel_transform_reduce_submitter
+{
+    template <typename... _Name>
+    using _KernelName = __reduce_kernel<_IsGPU{}, _Name...>;
+
+    template <typename _ExecutionPolicy, typename _Size, typename _ReduceOp, typename _TransformOp1,
+              typename _TransformOp2, typename _InitType,
+              oneapi::dpl::__internal::__enable_if_device_execution_policy<_ExecutionPolicy, int> = 0,
+              typename... _Ranges>
+    static auto
+    submit(_ExecutionPolicy&& __exec, _Size __n, ::std::uint16_t __work_group_size, _ReduceOp __reduce_op,
+           _TransformOp1 __transform_op1, _TransformOp2 __transform_op2, _InitType __init, _Ranges&&... __rngs)
+    {
+        using _Policy = typename ::std::decay<_ExecutionPolicy>::type;
```

Comment:
Could you include `<type_traits>` to make sure `std::decay` is available?

Reply:
There are other similar places in that header:
- `std::min` and `std::swap` require `<algorithm>`
- `std::uint16_t` requires `<cstdint>`

Reply:
I will include those headers.